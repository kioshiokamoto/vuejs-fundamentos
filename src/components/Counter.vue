<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }}</p>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>

  <div>
    <button @click="increment">+1</button>
    <button @click="decrement">-1</button>
  </div>
</template>

<script>
export default {
  name: "Counter",
  props: {
    title: String,
    start: {
      type: Number,
      // required: true,
      default: 0,
      validator(value) {
        return value >= 0;
      },
    },
  },
  data() {
    return {
      counter: this.start,
    };
  },

  methods: {
    getSquareValue() {
      return this.counter ** 2;
    },
    increment() {
      this.counter = this.counter + 1;
    },
    decrement() {
      this.counter = this.counter - 1;
    },
  },
  computed: {
    squareCounter() {
      // console.log("Computed square counter");
      return this.counter ** 2;
    },
    customTitle() {
      return this.title ? this.title : "Counter";
    },
  },
};
</script>

<style scoped>
button {
  background-color: #64db87;
  border-radius: 5px;
  border: none;
  color: #fff;
  cursor: pointer;
  margin: 0 5px;
  padding: 5px 15px;
  transition: 0.3s ease-in-out;
}
button:hover {
  background-color: #5aa67b;
}
</style>