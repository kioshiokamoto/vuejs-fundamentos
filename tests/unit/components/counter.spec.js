import { shallowMount, mount } from "@vue/test-utils";
import Counter from "@/components/Counter.vue";

describe("Counter Component", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(Counter);
    });

    test("Debe hacer match con el snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
    test("Debe renderizar etiqueta h2 con valor por defecto Counter", () => {
        expect(wrapper.find("h2").exists()).toBeTruthy();

        const h2 = wrapper.find("h2").text();

        expect(h2).toBe("Counter");
    });
    test("Debe incrementar contador", async () => {
        const increaseBtn = wrapper.find("button");

        await increaseBtn.trigger("click");

        const value = wrapper.find("p").text();

        expect(value).toBe("1");
    });
    test("Debe decrementar contador 2 veces", async () => {
        const increaseBtn = wrapper.findAll("button")[1];

        await increaseBtn.trigger("click");
        await increaseBtn.trigger("click");

        const value = wrapper.find("p").text();

        expect(value).toBe("-2");
    });
    test("Debe establecer valor por defecto", () => {
        const start = wrapper.props("start");

        const value = wrapper.find("p").text();

        expect(Number(value)).toBe(start);
    });
    test("Debe mostrar la prop title", () => {
        const title = "Contador";
        const wrapper = shallowMount(Counter, { props: { title } });
        const value = wrapper.find("h2").text();

        expect(value).toBe(title);
    });
});
