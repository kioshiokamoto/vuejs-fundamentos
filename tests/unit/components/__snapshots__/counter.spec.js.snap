// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Debe hacer match con el snapshot 1`] = `
<h2>Counter</h2>
<p>0</p>
<p>0 <sup>2</sup> = 0</p>
<div><button>+1</button><button>-1</button></div>
`;
